### PR TITLE
ROX-10993: Respect replaced resources within built-in scope checker.

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

We have added grouped resources within #1384 as a means to simplify the number of resources we currently expose. The main motivation behind this was to help the efforts within postgres SAC migration as well as help users simplify the creation of access scope, as the current resource model was confusing at times.

Now, we have added the concept of replacing resources for resources which are being deprecated as a field within each `ResourceMetadata`. We also adopted the `ResourceHelper`, which is most commonly used as a pattern to check for access to the resource within datastore implementations by conditionally creating a disjunction between two scope checkers: the one of the resource and, if given, the replacing resource.

What is still missing is the adoption of `ScopeCheckerCore` implementations to respect replaced resource.

Given the following example, which is a typical pattern within our code base:
```go
hasReadAccessToAPIToken := sac.WithGlobalAccessScopeChecker(context.Background(),
		sac.AllowFixedScopes(
			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
			sac.ResourceScopeKeys(resources.APIToken)))
```
We would expect that the `ScopeChecker` will correctly yield read access to both `APIToken` as well as its replacing resource `Integration`, but that's currently not the case.

Hence, we need to adopt the implementations of the `ScopeCheckerCore` to also take those into account.

Within this PR, the following has been done:
- Create helper function to check if a resource is contained within a permission map with the respective access.
- Adopt code to use the previously created helper function.
- Adopt the `BuiltInScopeChecker` to respect resource replacements.
- Add additional tests.

What has not been addressed within this PR and needs to be done going forward:
- Adopt `AllowFixedScopes` implementation.
- Adopt `ScopeCheckerCoreImpl` implementation.

## Testing Performed
- Unit tests (see CI).

### CODE REVIEW

**Review by rhybrillou — COMMENTED** [review]
Good job.
Please make sure the test descriptions match the test case data.

File: pkg/auth/permissions/resource_test.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import (
+	"testing"
+
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCheckResourceForAccess(t *testing.T) {
+	testResource := Resource("Test")
+	replacingResource := Resource("Test")
```
Comment (rhybrillou) [inline pkg/auth/permissions/resource_test.go:12]: Would it make sense to have a different resource name here ?

**Review by rhybrillou — APPROVED** [review]

File: pkg/auth/permissions/resource.go [hunk 2]
```diff
@@ -43,6 +45,11 @@ func (m ResourceMetadata) GetResource() Resource {
 
 // GetScope returns the resource scope for this metadata object.
 func (m ResourceMetadata) GetScope() ResourceScope {
+	// Replacing resources _may_ have a different ResourceScope than the initial resource.
+	// This way, clients don't have to deal with replacing resources when checking for scope.
+	if m.ReplacingResource != nil && m.ReplacingResource.GetScope() < m.Scope {
```
Comment (rukletsov) [inline pkg/auth/permissions/resource.go:50]: I hung here for a while. What this actually does is _broadens_ the scope, i.e., from namespace to cluster if the replacing resource has the broader one. If so, could you please adjust the comment to make this clear?

File: pkg/auth/permissions/resource.go [hunk 3]
```diff
@@ -71,3 +78,19 @@ func WithLegacyAuthForSAC(md ResourceMetadata, use bool) ResourceMetadata {
 	md.legacyAuthForSAC = &use
 	return md
 }
+
+// CheckResourceForAccess will verify whether the given ResourceMetadata is contained within the map
+// with at least the specified storage.Access.
+// Note: This will take replacing resources into account.
+func CheckResourceForAccess(resource ResourceMetadata, resourceAccessMap map[string]storage.Access,
```
Comment (rukletsov) [inline pkg/auth/permissions/resource.go:85]: Maybe take `ResourceMetadata` as a receiver and rename to `IsPermittedBy()`?

File: pkg/auth/permissions/resource.go (hunk 2 above)
Comment (dhaus67) [reply pkg/auth/permissions/resource.go:50]: I've made the changes within #1834 , let's move the discussion there, if there are any additional comments.

File: pkg/auth/permissions/resource.go (hunk 3 above)
Comment (dhaus67) [reply pkg/auth/permissions/resource.go:85]: I've made the changes within #1834 , let's move the discussion there, if there are any additional comments.

**Pull request merged**